### PR TITLE
Perf/Use readwrite lock with statesync

### PR DESCRIPTION
**Pull request by asdacap**
Includes #4915

- Statesync/healing currently have a global lock. 
- Meaning it's processing is limited to a single thread. 
- This change changes the lock to use read write lock so that part of it can run in parallel.
- Goerli state sync time reduce from 3 hours 30 minute to about 2 hours 30 minute.
- Mainnet healing time reduced significantly from 3 to 4 hours to under 1 hour (the snap sync was finished about a day earlier, so healing takes significantly more time than usual). 
  - This is because bottleneck in healing is during checking if a node is already saved which can now run in parallel. If a node is already saved, (like with snap sync) the call is several time longer than if the node does not exist since it actually load the data.
  - Additionally lower healing performance significantly increase healing time as it need to catch up with the network and every time state root change, its dependency graph need to be rebuild, which explain why some user finished healing very quickly while other take several hours.

Screenshot from 2022-11-21 10-02-06

## Changes:
- Use read write lock instead of plain lock.
- Add limit to reset root due to hint. This is because emptyish request is processed faster which without a global lock, will get re-send and appear much more often. 
- Modify handle trie node to pass unit test.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [X] Perf
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [X] No: Test from https://github.com/NethermindEth/nethermind/pull/4915 is relevant.

**Comments about testing , should you have some** (optional)

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by dceleda** [comment]
@kamilchodola could you confirm perf improvement by running more tests?

**Review by damian-orzechowski — COMMENTED** [review]
All looks good to me, but it would be good for someone else cast an eye - especially on 2 additonal test cases.

**Comment by asdacap** [comment]
@kamilchodola ran a few test.

Goerli on EPYC 7601 shows clear improvement in state sync time.

Screenshot from 2023-01-02 17-02-52

Mainnet on EPYC 7601 shows largely improved state sync time.

Screenshot from 2023-01-02 17-02-33

On 7713, goerli does not show clear improvement (all state sync time is less than 2 minute).
Did not get scheduled on 7713 for this branch, so did not get matching CPU.

**Comment by kamilchodola** [comment]
@dceleda @asdacap @MarekM25 @damian-orzechowski 
Overall I must say I can see some improvements in averaged results especially on goerli - for mainnet need more time since Linode wasn't lucky enough to gave me exact same setup for master and this branch so will rerun but seems really promising at the moment.

**Review by kamilchodola — APPROVED** [review]

**Review by damian-orzechowski — APPROVED** [review]

File: src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs [hunk 1]
```diff
@@ -133,7 +138,8 @@ public SyncResponseHandlingResult HandleResponse(StateSyncBatch? batch)
 
             try
             {
-                lock (_handleWatch)
+                _pendingRequestsLock.EnterReadLock();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs:141]: This is confusing. It enters `ReadLock` but changes things. I expect the assumption is that those changes are thread-safe. 
1. Looks ok:
- `_pendingRequests` as it is `ConcurrentDictionary`.
- `_lastReview` 
2. Looks like a potential access issue:
- `_pendingItems.RecalculatePriorities` 
- `_hintsToResetRoot`
- `_branchProgress` in `AddNodeToPending`
- `_branchProgress` in `SaveNode`
- `_branchProgress` in `HandleTrieNode`
- `_data`
- `_handleWatch` (not Interlocked)
3. Other questions:
- Do we need to call `StoreProgressInDb` every time?

File: src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs [hunk 2]
```diff
@@ -635,12 +714,22 @@ private void VerifyPostSyncCleanUp()
         private void StoreProgressInDb()
         {
             byte[] serializedData = _data.Serialize();
-            lock (_stateDbLock)
+            _stateDbLock.EnterWriteLock();
+            try
             {
-                lock (_codeDbLock)
+                _codeDbLock.EnterWriteLock();
+                try
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs:721]: Do we need both locks here?

**Review by LukaszRozmej — COMMENTED** [review]
Can you explain `dependentItem.Counter` changes?

File: src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs:721]: Not sure... previously it does.

File: src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs:141]: Seems like I missed a comment here. I guess it was skipped on the various rebase I did?. Its a readwritelock, but not really a read-write operation. It just need to have exclusive access during the write (as in cleaning up the whole syns state), while allowing concurrent read (during evrything else). Specifically, it protects the five field below it. I guess a better name would be `_syncStateLock`.

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs:141]: ok but now everything under `Read` lock is accessed concurrently and I mentioned in point `2` all the fields that might break on that access.

**Comment by asdacap** [comment]
As mentioned in the comment, the counter is set to its max value first as during the loop of the node's children, its possible that its children is saved at that time which decrement the counter. This might cause the counter to be zero and the node to be saved prematurely.

File: src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs]: Done

Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/FastSync/TreeSync.cs]: Done

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**